arcgis-ios-sdk-samples/Display information/Filter by time extent/FilterByTimeExtentViewController.swift

```diff
@@ -0,0 +1,128 @@
+import ArcGIS
+import ArcGISToolkit
+
+class FilterByTimeExtentViewController: UIViewController {
+    /// The map view.
+    @IBOutlet var mapView: AGSMapView! {
+        didSet {
+            mapView.map = AGSMap(basemapStyle: .arcGISTopographic)
+            // Set the map view's viewpoint.
+            let center = AGSPoint(x: -58.495293, y: 29.979774, spatialReference: .wgs84())
+            mapView.setViewpoint(AGSViewpoint(center: center, scale: 147_914_381.897889))
+        }
+    }
+    
+    /// The date formatter.
+    static let formatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "yyyy/MM/dd HH:mm"
+        return formatter
+    }()
+    
+    /// The default date and time for the starting thumb.
+    let currentStartTime = formatter.date(from: "2005/10/01 05:00")!
+    /// The default date and time for the ending thumb.
+    let currentEndTime = formatter.date(from: "2005/10/31 05:00")!
+    
+    /// The feature layer tracking hurricanes in 2005.
+    let featureLayer = AGSFeatureLayer(
+        item: AGSPortalItem(
+            portal: .arcGISOnline(withLoginRequired: false),
+            itemID: "49925d814d7e40fb8fa64864ef62d55e"
+        ),
+        layerID: 0
+    )
+    
+    /// The time slider from the ArcGIS toolkit.
+    let timeSlider = TimeSlider()
+    
+    /// Configure the time slider's attributes and position.
+    func setupTimeSlider() {
+        // Configure time slider.
+        timeSlider.labelMode = .thumbs
+        timeSlider.addTarget(self, action: #selector(timeSliderValueChanged(timeSlider:)), for: .valueChanged)
+        // Add the time slider to the view.
+        view.addSubview(timeSlider)
+        
+        // Add constraints to position the slider.
+        let margin: CGFloat = 10.0
+        timeSlider.translatesAutoresizingMaskIntoConstraints = false
+        // Assign the constraints.
+        let constraints = [
+            timeSlider.bottomAnchor.constraint(equalTo: mapView.attributionTopAnchor, constant: -margin),
+            timeSlider.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: margin),
+            timeSlider.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -margin)
+        ]
+        // Activate the constraints.
+        NSLayoutConstraint.activate(constraints)
+    }
+    
+    /// Initialize the time slider's steps.
+    func initializeTimeStepsFromQuery() {
+        featureLayer.load { [unowned self] error in
+            guard error == nil else { return }
+            populateFeaturesWithQuery { timeExtent in
+                timeSlider.initializeTimeSteps(timeStepCount: 200, fullExtent: timeExtent) { _ in
+                    // Show the time slider.
+                    timeSlider.currentExtent = AGSTimeExtent(startTime: currentStartTime, endTime: currentEndTime)
+                    
+                    mapView.map!.operationalLayers.add(featureLayer)
```

Comment:
It's better to not force-unwrap this.  The `mapView` *should* always have a map, but if it doesn't the following won't crash.

```suggestion
                    mapView.map?.operationalLayers.add(featureLayer)
```

Reply:
While we don't want users to see any crashes, they can be a good thing in development as they indicate that something is wrong. In this case, crashing would indicate that the map view isn't configured directly, which I would think is a good thing as we would want to know that as soon as possible. Using optional chaining here might seem like it is benefitting the user by eliminating a potential crash, but that may not be the case. Being overly defensive could actually end up harming the user if it prevents issues from being detected at development time. Therefore I would say that force unwrapping here is fine. 

It isn't unlike IB outlets. We use implicitly unwrapped optionals because our assumption is that they will be non-optional when used. We could conditionally unwrap them every time, but that would be tedious and could actually end up being harmful as we wouldn't get immediate feedback if a connection in the storyboard/xib breaks. Does that make sense?

Reply:
I agree with Phil, fail purposely is why the ❗ used here.

But I think Viv is still making changes to the code and eventually this might not be needed. If we are not expecting to have the feature layer finish loading before setting up the time slider, then it can be added to the map's `operationalLayers` in mapView `didSet`, which would avoid this force unwrap.

Reply:
Ok, fine with me.